src/common.c

```diff
@@ -29,146 +29,158 @@ static int initialized = 0;
+        ++q;
+    return !*q && !(((p-q))&1);
 }
 
 int ishexn(const char *q, int n)
 {
-	const char *p=q;
-	while (atoi16[ARCH_INDEX(*q)] != 0x7F)
-		++q;
-	return (q-p) >= n;
+    const char *p=q;
+    while (atoi16[ARCH_INDEX(*q)] != 0x7F)
+        ++q;
+    return (q-p) >= n;
 }
 int ishexucn(const char *q, int n)
 {
-	const char *p=q;
-	while (atoi16u[ARCH_INDEX(*q)] != 0x7F)
-		++q;
-	return (q-p) >= n;
+    const char *p=q;
+    while (atoi16u[ARCH_INDEX(*q)] != 0x7F)
+        ++q;
+    return (q-p) >= n;
 }
 int ishexlcn(const char *q, int n)
 {
-	const char *p=q;
-	while (atoi16l[ARCH_INDEX(*q)] != 0x7F)
-		++q;
-	return (q-p) >= n;
+    const char *p=q;
+    while (atoi16l[ARCH_INDEX(*q)] != 0x7F)
+        ++q;
+    return (q-p) >= n;
 }
 
 int ishexuc_oddOK(const char *q) {
-	while (atoi16[ARCH_INDEX(*q)] != 0x7F) {
-		if (*q >= 'a' && *q <= 'f') return 0;
-		++q;
-	}
-	return !*q ;
+    while (atoi16[ARCH_INDEX(*q)] != 0x7F) {
+        if (*q >= 'a' && *q <= 'f') return 0;
+        ++q;
+    }
+    return !*q ;
 }
 int ishexlc_oddOK(const char *q) {
-	while (atoi16[ARCH_INDEX(*q)] != 0x7F) {
-		if (*q >= 'A' && *q <= 'F') return 0;
-		++q;
-	}
-	return !*q ;
+    while (atoi16[ARCH_INDEX(*q)] != 0x7F) {
+        if (*q >= 'A' && *q <= 'F') return 0;
+        ++q;
+    }
+    return !*q ;
 }
 
 static MAYBE_INLINE size_t _hexlen(const char *q, unsigned char dic[0x100], int *extra_chars)
 {
-	const char *s = q;
-	size_t len = strlen(q);
-	if (len&1) --len;
+    const char *s = q;
+    size_t len = strlen(q);
+    if (len&1) --len;
 
-	while (dic[ARCH_INDEX(*q)] != 0x7F)
-		++q;
-	if ((size_t)(q - s)&1) --q;
-	if (extra_chars)
-		*extra_chars = (*q != 0);
-	return (q - s);
+    while (dic[ARCH_INDEX(*q)] != 0x7F)
+        ++q;
+    if ((size_t)(q - s)&1) --q;
+    if (extra_chars)
+        *extra_chars = (*q != 0);
+    return (q - s);
 }
 size_t hexlen(const char *q, int *extra_chars)
 {
-	return _hexlen(q, atoi16, extra_chars);
+    return _hexlen(q, atoi16, extra_chars);
 }
 size_t hexlenu(const char *q, int *extra_chars)
 {
-	return _hexlen(q, atoi16u, extra_chars);
+    return _hexlen(q, atoi16u, extra_chars);
 }
 size_t hexlenl(const char *q, int *extra_chars)
 {
-	return _hexlen(q, atoi16l, extra_chars);
+    return _hexlen(q, atoi16l, extra_chars);
 }
-
 int isdec(const char *q)
 {
-	char buf[24];
-	int x = atoi(q);
-	sprintf(buf, "%d", x);
-	return !strcmp(q,buf) && *q != '-';
+    const char *imax = "2147483647";
+    const char *p = q;
+    do {
+        if (*p < '0' || *p > '9' || p - q >= 10)
+            return 0;
+    } while (*++p);
+
+    return strcmp(q,imax) <= 0;
 }
 int isdec_negok(const char *q)
 {
-	char buf[24];
-	int x = atoi(q);
-	sprintf(buf, "%d", x);
-	return !strcmp(q,buf);
+
+    const char *imin = "2147483648";
+    const char *p = (*q == '-')?++q:q;
+    do {
+        if (*p < '0' || *p > '9' || p - q >= 10)
+            return 0;
+    } while (*++p);
+
+    return(strcmp(q,imin) <= 0);
```

Comment:
This will wrongly allow the positive value 2147483648, which is one higher than INT_MAX. I think it's best to write a shared static function that will contain this loop and the `strcmp()` call, and will accept pointers to the string to be checked and to the string containing the max number. We'll call this function differently from all three `isdec*()`, and in particular `isdec_negok()` will call it differently based on whether the original string started with a minus sign (skip the minus and use the one higher max string) or not (just `return isdec(...)` instead, which will also use the shared function internally).

Reply:
@solardiz Thanks for the feedback. I just though about it when i was in bed, but was too late. I new the comparison at was wrong. Thanks. I like your suggestion to minimize code duplication. I will do that.

@jfoug  in `isdec_negok()` when the 10 digits number has option '-' then it should be in that range of  <= 0x80000000 (-2147483648 in 2's complement) right? which makes cases like this true `-1147483648 <= -2147483648` right?